### ISSUE TITLE
NaNs in parsed telemetry JSON cause bad things to happen.

### ISSUE THREAD

**Issue by danielrichman** [ukhas/habitat#248]
Background
- JSON spec doesn't allow NaN (e.g., {"blah": NaN}) however it is occasionally seen since Javascript supports this and parsing  using javascript eval() would work
- python json.dumps and json.loads support it. Switched on by default. See "allow_nan" of json.dumps; switching that off causes NaNs to become ValueErrors.
- also: -inf/-Infinity and inf/Infinity (same deal. allow_nan=False also disables these).

Observations.
- parser appears to crash after parsing
  
  ```
  [2012-08-29 12:24:55,894] INFO habitat.parser MainThread: Parsing [ascii] '$$LUMIA,2,13:24:50,47.644830,-122.141197,NaN,NaN,1*671E\n' (4eb38eae5c15b9dbd05ef8d996e56fb5945ae1356c6e7bfda655e140be4fe1a9)
  [2012-08-29 12:24:55,912] DEBUG habitat.parser MainThread: Parsed data: {"ALTITUDE": NaN, "_parsed": true, "DATA_CONNECTION": 1, "_sentence": "$$LUMIA,2,13:24:50,47.644830,-122.141197,NaN,NaN,1*671E\n", "longitude": -122.141197, "sentence_id": 2, "_flight": "33acb4a3a56276f20d26cf7e051333b3", "_protocol": "UKHAS", "time": {"second": 50, "minute": 24, "hour": 13}, "latitude": 47.64483, "SPEED": NaN, "payload": "LUMIA"}
  [2012-08-29 12:24:55,912] INFO habitat.parser MainThread: UKHAS parsed data from LUMIA successfully
  ```
- then, it goes silent - nothing written to log file. No 100% cpu or anything. Any more uploaded sentences - no response.
- this parser was not (and is not) using immortal changes. However it did not log or print an exception to stderr and it did not die. Date modifited on /var/log/supervisor/blah is days ago.
- strace shows that it is still reading the heartbeat and changes from _changes:
  
  ```
  recv(4, "1\r\n\n\r\n", 8192, 0)         = 6
  recv(4, "1\r\n\n\r\n", 8192, 0)         = 6
  recv(4, "1\r\n\n\r\n", 8192, 0)         = 6
  recv(4, "1b1\r\n{\"seq\":1822408,\"id\":\"99d984"..., 8192, 0) = 440
  recv(4, "1\r\n\n\r\n", 8192, 0)         = 6
  recv(4, "1\r\n\n\r\n", 8192, 0)         = 6
  ```
- however, no response. Does not attempt to parse anything or write anything to the log.
- in couchdb log from exactly the same time:
  
  ```
  [Wed, 29 Aug 2012 12:24:55 GMT] [info] [<0.4492.7>] 80.6.136.243 - - PUT /habitat/4eb38eae5c15b9dbd05ef8d996e56fb5945ae1356c6e7bfda655e140be4fe1a9 201
  [Wed, 29 Aug 2012 12:24:55 GMT] [info] [<0.8229.7>] 127.0.0.1 - - GET /habitat/_design/habitat/_view/payload_config?startkey=%5B%22LUMIA%22%2C+1346243090%5D&limit=1&include_docs=true 200
  [Wed, 29 Aug 2012 12:24:55 GMT] [info] [<0.8229.7>] 127.0.0.1 - - GET /habitat/4eb38eae5c15b9dbd05ef8d996e56fb5945ae1356c6e7bfda655e140be4fe1a9 200
  [Wed, 29 Aug 2012 12:24:55 GMT] [error] [<0.8229.7>] attempted upload of invalid JSON (set log_level to debug to log it)
  [Wed, 29 Aug 2012 12:24:55 GMT] [info] [<0.8229.7>] 127.0.0.1 - - PUT /habitat/4eb38eae5c15b9dbd05ef8d996e56fb5945ae1356c6e7bfda655e140be4fe1a9 400
  ```
- this is what happens if you try save such a doc:
  
  ```
  >>> d.save_doc({"type": "test", "value": float('nan')})
  Traceback (most recent call last):
    File "<stdin>", line 1, in <module>
    File "/home/habitat/venv/local/lib/python2.7/site-packages/couchdbkit/client.py", line 498, in save_doc
      doc1.update({'_id': res['id'], '_rev': res['rev']})
  TypeError: string indices must be integers, not str
  ```

**Comment by danielrichman** [comment]
```
>>> def callback(*args):
...     print "In callback"
...     try:
...         res = d.save_doc({"type": "test", "value": float('nan')})
...     except:
...         print "A traceback!"
...         traceback.print_exc()
...         print "Re raising"
...         raise
...     else:
...         print "The result", repr(res)
>>>
>>>
>>> c.wait(callback, since=1822542, include_docs=True, heartbeat=1000)
In callback
A traceback!
Traceback (most recent call last):
  File "<stdin>", line 4, in callback
  File "/home/daniel/.python/lib/python2.6/site-packages/couchdbkit/client.py", line 498, in save_doc
    doc1.update({'_id': res['id'], '_rev': res['rev']})
TypeError: string indices must be integers, not str
Re raising
```

Now, it hangs. Ctrl-C reveals where it's stuck:

```
^CTraceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/daniel/.python/lib/python2.6/site-packages/couchdbkit/consumer/__init__.py", line 103, in wait
    return self._consumer.wait(cb, **params)
  File "/home/daniel/.python/lib/python2.6/site-packages/couchdbkit/consumer/sync.py", line 56, in wait
    break
  File "/home/daniel/.python/lib/python2.6/site-packages/restkit/wrappers.py", line 161, in __exit__
    self.close()
  File "/home/daniel/.python/lib/python2.6/site-packages/restkit/wrappers.py", line 169, in close
    self.body.read()
  File "/usr/lib/python2.6/io.py", line 940, in read
    return self._read_unlocked(n)
  File "/usr/lib/python2.6/io.py", line 955, in _read_unlocked
    chunk = self.raw.read()
  File "/usr/lib/python2.6/io.py", line 589, in read
    return self.readall()
  File "/usr/lib/python2.6/io.py", line 599, in readall
    data = self.read(DEFAULT_BUFFER_SIZE)
  File "/usr/lib/python2.6/io.py", line 591, in read
    n = self.readinto(b)
  File "/home/daniel/.python/lib/python2.6/site-packages/http_parser/reader.py", line 60, in readinto
    recved = self.http_stream.stream.readinto(buf)
  File "/home/daniel/.python/lib/python2.6/site-packages/http_parser/reader.py", line 150, in readinto
    return _readinto(self._sock, b)
  File "/home/daniel/.python/lib/python2.6/site-packages/http_parser/reader.py", line 25, in _readinto
    buf = sock.recv(l)
KeyboardInterrupt
```

Note "in __ exit __" - called when leaving a "with" block (possibly via an exception).

Guess this is two bugs in couchdbkit/restkit: a) it tries to save NaNs, b) if an exception is thrown in a continuous changes callback it just hangs rather than error exiting. immortal_changes should "solve"/cover up (b), since it logs and absorbs exceptions raised by the callback. (a) isn't that much of a bug, since if they did pass allow_nan=False then it would simply raise a ValueError instead of the above odd TypeError.

Solution is probably to just have base.ascii_float reject nan,inf,-inf.

**Comment by danielrichman** [comment]
I checked out the latest version of couchdbkit. If you try to save nans it produces this far nicer error message:

``` python
>>> db.save_doc({"type": "test", "value": float('nan')})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/daniel/temp/src/couchdbkit/couchdbkit/client.py", line 508, in save_doc
    res = self.res.post(payload=doc1, **params).json_body
  File "/home/daniel/temp/lib/python2.6/site-packages/restkit/resource.py", line 144, in post
    headers=headers, params_dict=params_dict, **params)
  File "/home/daniel/temp/src/couchdbkit/couchdbkit/resource.py", line 111, in request
    payload=payload, headers=headers, **params)
  File "/home/daniel/temp/lib/python2.6/site-packages/restkit/resource.py", line 210, in request
    response=resp)
restkit.errors.RequestFailed: {"error":"bad_request","reason":"invalid_json"}
```

Admittedly it has just sent invalid json to the server which it probably shouldn't do... but I don't think this warrants a bug report to couchdbkit (?)

Here's my issue for (b)
https://github.com/benoitc/couchdbkit/issues/147

**Comment by adamgreig** [comment]
Closed by #249
